### ISSUE TITLE
Sorting or not sorting candidates

### ISSUE THREAD

**Issue by DamienCassou** [minad/vertico#228]
I think there are basically 2 kinds of usages for completing-read that I think could be handled differently:

- dynamic or long list of candidates
- short list of candidates that are always the same

In the first case, sorting candidates by, e.g., history makes sense.

In the second case, I think sorting candidates is problematic because it prevents users from learning the position of the candidates by heart. For example, when opening a picture file from dired, I can choose between "gimp", "eog" and a few more. I configured "eog" to be at the beginning because this is what I want by default and I learned I can just press RET to get "eog". If I open "gimp" one day, I don't want "gimp" to be the new first candidate.

For now, I'm using vertico-multiform to deactivate sorting for commands that make sense:

```emacs
(setq vertico-multiform-commands
          '(;; don't sort candidates of commands with predictable
            ;; candidates because my brain wants to learn their
            ;; position:
            (dired-do-async-shell-command (vertico-sort-function . nil))
            (tmr (vertico-sort-function . nil))
            (prodigy-browse (vertico-sort-function . nil))))
```

I'm wondering if vertico couldn't get an option containing such command names that would disable sorting for them.

**Comment by minad** [comment]
> In the second case, I think sorting candidates is problematic because it prevents users from learning the position of the candidates by heart.

This is true. Therefore `completing-read` offers the possibility to disable sorting. You can set the `display-sort-function` and the `cycle-sort-function` in the completion metadata to `identity`.

> I'm wondering if vertico couldn't get an option containing such command names that would disable sorting for them.

It does - `vertico-multiform-commands`. Maybe that's not the most convenient approach but you could easily write a helper function and add it to your config if the problem occurs often for you.

Please let me know if this answers your question!

**Comment by DamienCassou** [comment]
> Therefore completing-read offers the possibility to disable sorting

This makes perfect sense. I will open PRs for the packages defining these commands.

> It does - vertico-multiform-commands.

Yes and no. I agree disabling sorting is doable by the end-user. What I had in mind was more a pre-populated list of commands we know shouldn't sort candidates. Nevertheless, now that I know it can be fixed upstream, I think I should do that instead.

**Comment by minad** [comment]
> What I had in mind was more a pre-populated list of commands we know shouldn't sort candidates. Nevertheless, now that I know it can be fixed upstream, I think I should do that instead.

Understood. In my opinion such things are better solved upstream such that all completion UIs are informed about the desired sorting. For the cases where we want Vertico-specific overrides, vertico-multiform is good enough - the mechanism is highly flexible. Maintaining preconfigured lists however is out of scope of Vertico.

**Comment by minad** [comment]
The default is always moved to the top by Vertico. This is intentional and reasonable since RET should return this candidate. If you don't like this, don't specify a default value.

**Comment by DamienCassou** [comment]
Excellent. Thank you very much for your explanation. I think I'm good now and submitted patches: https://lists.sr.ht/~protesilaos/tmr/%3C87o7zwhcqn.fsf%40cassou.me%3E.
